db.go

```diff
@@ -501,25 +528,35 @@ func (db *DB) syncWALToLTX(ctx context.Context, ltxFilename string) error {
 		return err
 	}
 
-	// If salt matches the salt in the LTX file & the WAL size exceeds the
-	// offset in the LTX file then we need to truncate the WAL to match.
-	// This occurs before the checkpoint() call so we don't replay too much.
+	// If WAL salt doesn't match the LTX WAL salt then the WAL has been
+	// restarted and we need to remove it. We are just renaming it for now so
+	// we can debug in case this happens.
 	salt1 := binary.BigEndian.Uint32(hdr[16:])
 	salt2 := binary.BigEndian.Uint32(hdr[20:])
 	if salt1 != r.Header().WALSalt1 || salt2 != r.Header().WALSalt2 {
-		log.Printf("wal salt mismatch on %q, skipping sync with ltx", db.name)
-		return nil
-	} else if fi.Size() <= ltxWALSize {
-		log.Printf("wal size in sync on %q, skipping truncation", db.name)
+		log.Printf("wal salt mismatch on %q, removing wal", db.name)
+		if err := os.Rename(db.WALPath(), db.WALPath()+".removed"); err != nil {
+			return fmt.Errorf("rename wal file with salt mismatch: %w", err)
+		}
 		return nil
 	}
 
+	// If the salt matches then we need to make sure we are at least up to the
+	// start of the last LTX transaction.
+	if fi.Size() < r.Header().WALOffset {
+		return fmt.Errorf("short wal size (%d bytes) for %q, last ltx offset at %d bytes", fi.Size(), db.name, r.Header().WALOffset)
+	}
+
```

Comment:
If the WAL file is shorter than the initial offset of the last LTX file then we're missing a transaction. Go ahead and error since we can't recover from this.

Technically, we could use earlier LTX files to sync up the position but we are performing `fsync(2)` calls immediately after WAL commit so we should never be more than 1 LTX file off.